File: src/main/java/me/taubsie/carrylogs/application/command/commands/CalcPriceCommand.java
package me.taubsie.carrylogs.application.command.commands;

import me.taubsie.carrylogs.application.command.Command;
import me.taubsie.carrylogs.application.command.CommandParameters;
import me.taubsie.carrylogs.application.connection.DungeonHubConnection;
import me.taubsie.carrylogs.application.enums.*;
import me.taubsie.carrylogs.application.exceptions.CommandExecutionException;
import me.taubsie.carrylogs.application.exceptions.InvalidOptionException;
import me.taubsie.carrylogs.application.service.ApplicationService;
import me.taubsie.dungeonhub.common.CarryDifficulty;
import me.taubsie.dungeonhub.common.CarryTier;
import org.javacord.api.entity.DiscordEntity;
import org.javacord.api.entity.channel.Categorizable;
import org.javacord.api.entity.channel.Channel;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@CommandParameters(name = "calc-price",
        description = "Calculate the price for some amount of carries.")
public class CalcPriceCommand extends Command {
    public static long calculatePrice(CarryDifficulty carryDifficulty, long amount) {
        Optional<Integer> bulkPrice = carryDifficulty.getBulkPrice();
        Optional<Integer> bulkAmount = carryDifficulty.getBulkAmount();

        if (bulkPrice.isPresent() && bulkAmount.isPresent() && bulkAmount.get() <= amount) {
            return bulkPrice.get() * amount;
        }

        return carryDifficulty.getPrice() * amount;
    }

    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        Server server = getServer();

        long amount = getLongOption("amount");

        Optional<CarryTier> carryTier = slashCommandCreateEvent.getSlashCommandInteraction().getChannel()
                .flatMap(Channel::asCategorizable)
                .flatMap(Categorizable::getCategory)
                .map(DiscordEntity::getId)
                .flatMap(id -> DungeonHubConnection.getInstance().getCarryTierFromCategory(server.getId(), id));

        try {
            Optional<CarryTier> previousCarryTier = carryTier;

            carryTier = DungeonHubConnection.getInstance()
                    .loadCarryType(server.getId(), getStringOption("carry-type"))
                    .flatMap(carryType -> DungeonHubConnection.getInstance().loadCarryTier(carryType,
                            getStringOption("carry-tier")))
                    .or(() -> previousCarryTier);
        }
        catch (CommandExecutionException commandExecutionException) {
            //ignored since the checking happens later
        }

        if (carryTier.isEmpty()) {
            throw new InvalidOptionException("carry-tier", "Please either use this in a carry ticket or supply a " +
                    "carry type and carry tier.");
        }

        Optional<CarryDifficulty> carryDifficulty = DungeonHubConnection.getInstance()
                .loadCarryDifficulty(carryTier.get(), getStringOption("carry-difficulty"));

        if (carryDifficulty.isEmpty()) {
            throw new InvalidOptionException("carry-difficulty");
        }

        long price = calculatePrice(carryDifficulty.get(), amount);

        if (price < 0) {
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Something went wrong.. The calculated price (" + price + ") is negative?";
                }
            };
        }

        String priceText = price != 0
                ? ApplicationService.getInstance().makeNumberReadable(price) + " coins"
                : "Free";

        EmbedBuilder embed = ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.INFORMATION.getColor())
                .setTitle("Carry-Price")
                .addInlineField("Type", carryTier.get().getDisplayName() + " | " + carryDifficulty.get().getDisplayName())
                .addInlineField("Amount", String.valueOf(amount))
                .addInlineField("Price", priceText);

        carryDifficulty.flatMap(CarryDifficulty::getThumbnailUrl).ifPresent(embed::setThumbnail);

        respond(embed);
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption amountOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.LONG)
                .setName("amount")
                .setDescription("The amount of carries you want.")
                .setLongMaxValue(200)
                .setLongMinValue(1L)
                .setRequired(true)
                .build();

        return Arrays.asList(CarryTypeCommand.getCarryTypeOption(),
                CarryTierCommand.getCarryTierOption(),
                CarryDifficultyCommand.getCarryDifficultyOption(),
                amountOption);
    }
}

File: src/main/kotlin/net/dungeonhub/application/commands/CalcPriceCommand.kt
package net.dungeonhub.application.commands

import dev.kord.core.behavior.channel.asChannelOfOrNull
import dev.kord.core.entity.channel.CategorizableChannel
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.converters.impl.long
import dev.kordex.core.commands.converters.impl.string
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import dev.kordex.core.utils.getLocale
import net.dungeonhub.application.enums.EmbedColor
import net.dungeonhub.application.exceptions.CommandExecutionException
import net.dungeonhub.application.exceptions.InvalidOptionException
import net.dungeonhub.application.loader.LoadExtension
import net.dungeonhub.application.service.ApplicationService
import net.dungeonhub.application.service.AutoCompletionService
import net.dungeonhub.auth.AuthenticationConnection
import net.dungeonhub.connection.CarryDifficultyConnection
import net.dungeonhub.connection.CarryTierConnection
import net.dungeonhub.connection.CarryTypeConnection
import net.dungeonhub.connection.DiscordServerConnection
import net.dungeonhub.i18n.Translations

/**
 * Command to calculate the price for some amount of carries.
 * This command can be used in a carry ticket channel to automatically get the price for the current carry tier.
 * If used outside a carry ticket channel, the user has to supply the carry type and carry tier.
 * The command will then calculate the price for the given amount of carries.
 */
@LoadExtension
class CalcPriceCommand : Extension() {
    override val name = "calc-price-command"

    override suspend fun setup() {
        publicSlashCommand(::CalcPriceArguments) {
            name = Translations.Command.CalcPrice.name
            description = Translations.Command.CalcPrice.description
            allowInDms = false

            action {
                respond {
                    var carryTier = channel.asChannelOfOrNull<CategorizableChannel>()
                        ?.categoryId
                        ?.let { id ->
                            DiscordServerConnection.authenticated(AuthenticationConnection).getCarryTierFromCategory(
                                guild!!.id.value.toLong(),
                                id.value.toLong()
                            )
                        }

                    val previousCarryTier = carryTier

                    carryTier = CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                        .getByIdentifier(arguments.carryType)
                        ?.let { carryType ->
                            CarryTierConnection[carryType].authenticated()
                                .getByIdentifier(arguments.carryTier)
                                ?: previousCarryTier
                        }

                    if (carryTier == null) {
                        throw InvalidOptionException(
                            "carry-tier", "Please either use this in a carry ticket or supply a " +
                                    "carry type and carry tier."
                        )
                    }

                    val carryDifficulty = CarryDifficultyConnection[carryTier].authenticated()
                        .getByIdentifier(arguments.carryDifficulty)
                        ?: throw InvalidOptionException("carry-difficulty")

                    val price = ApplicationService.calculatePrice(carryDifficulty, arguments.amount)
                    val pricePerCarry = ApplicationService.calculatePricePerCarry(carryDifficulty, arguments.amount)

                    if (price < 0) {
                        throw CommandExecutionException("Something went wrong.. The calculated price ($price) is negative?")
                    }

                    val priceText = if (price != 0L) "${ApplicationService.makeNumberReadable(price)} (${
                        ApplicationService.makeNumberReadable(pricePerCarry)
                    }) coins" else "Free"

                    val embed = ApplicationService.embed
                    embed.color = EmbedColor.Information.color
                    embed.title =
                        Translations.Command.CalcPrice.Response.title.withLocale(event.getLocale()).translate()
                    embed.field(
                        Translations.Command.CalcPrice.Response.Fields.type.translateLocale(event.getLocale()),
                        true
                    ) {
                        carryTier.displayName + " | " + carryDifficulty.displayName
                    }
                    embed.field(
                        Translations.Command.CalcPrice.Response.Fields.amount.translateLocale(event.getLocale()),
                        true
                    ) { arguments.amount.toString() }
                    embed.field(
                        Translations.Command.CalcPrice.Response.Fields.price.translateLocale(event.getLocale()),
                        true
                    ) { priceText }

                    carryDifficulty.thumbnailUrl?.let { embed.thumbnail { url = it } }

                    embeds = mutableListOf(embed)
                }
            }
        }
    }

    class CalcPriceArguments : Arguments() {
        val carryType by string {
            name = Translations.CommonArguments.CarryType.name
            description = Translations.CommonArguments.CarryType.description
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val carryTier by string {
            name = Translations.CommonArguments.CarryTier.name
            description = Translations.CommonArguments.CarryTier.description
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryTier
        }

        val carryDifficulty by string {
            name = Translations.CommonArguments.CarryDifficulty.name
            description = Translations.CommonArguments.CarryDifficulty.description
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryDifficulty
        }

        val amount by long {
            name = Translations.Command.CalcPrice.Arguments.Amount.name
            description = Translations.Command.CalcPrice.Arguments.Amount.description
            maxValue = 200
            minValue = 1
        }
    }
}